include/pybind11/pybind11.h

```diff
@@ -2562,7 +2597,7 @@ void implicitly_convertible() {
         ~set_flag() { flag = false; }
     };
     auto implicit_caster = [](PyObject *obj, PyTypeObject *type) -> PyObject * {
-        static bool currently_used = false;
+        static thread_local bool currently_used = false;
```

Comment:
hmm, this will have a non-negligible cost. You may at least want to #ifdef-gate it to the free-threaded case.